### DISCUSSION TITLE
CommandPaletteActionProvider - how to group entries?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#550] [Q&A] [answered]
Is there a way to group actions in a `CommandPaletteActionProvider`? 
Currently my palette shows all actions ordered alphanumeric:

![image](https://user-images.githubusercontent.com/1339124/154796801-86297c73-d8a7-4219-8c4c-0a192d4ac60a.png)

In videos I have seen that it should be possible. But I did not figure out how to realize this.

My code is more or less copied from the worklfow example and looks like this:

```
      actions.addAll(Sets.newHashSet(
         new LabeledAction("Create Manual Task",
            Lists.newArrayList(new CreateNodeOperation(ModelTypes.MANUAL_TASK, lastMousePosition.orElse(point(0, 0)),
               "fa-plus-square"))),
         new LabeledAction("Create User Task",
            Lists.newArrayList(new CreateNodeOperation(ModelTypes.USER_TASK, lastMousePosition.orElse(point(0, 0)),
               "fa-plus-square"))),
         ....
      ));
```
It seems that the interface CommandPaletteActionProvider did not provide any grouping?

Thanks for any hints or help.

**Comment by planger** [comment, marked as answer]
The command palette action provider is used to specify the actions shown in the palette that opens up on <kbd>Ctrl</kbd>+<kbd>Space</kbd>. In the GLSP code, the palette shown in your screenshot is called _tool palette_ and its items are provided by an implementation of the `ToolPaletteItemProvider`.

The [`DefaultToolPaletteItemProvider`](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/internal/toolpalette/DefaultToolPaletteItemProvider.java) collects all registered instances of `CreateNodeOperationHandler` and `CreateEdgeOperationHandler` and shows them in the group _Nodes_ and _Edges_ respectively. This is what you see in your screenshot.

If you want to provide custom grouping, you'll have to bind a custom `ToolPaletteItemProvider`.
```java
   @Override
   protected Class<? extends ToolPaletteItemProvider> bindToolPaletteItemProvider() {
      return MyToolPaletteItemProvider.class;
   }
```

To create groups, use the following utils in the implementation of `getItems` in your provider (see also the implementation of  [`DefaultToolPaletteItemProvider`](https://github.com/eclipse-glsp/glsp-server/blob/master/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/internal/toolpalette/DefaultToolPaletteItemProvider.java)).
```java
      return Lists.newArrayList(
         PaletteItem.createPaletteGroup("node-group", "Nodes", nodes, "symbol-property", "A"),
         PaletteItem.createPaletteGroup("edge-group", "Edges", edges, "symbol-property", "B"));
```

In addition, it is always possible to customize the entire tool palette on the client by rebinding your own implementation of the [tool palette module](https://github.com/eclipse-glsp/glsp-client/tree/master/packages/client/src/features/tool-palette).

**Reply by rsoika** [reply]
Thanks for clarification.
I have created a short doc page with the information you provided and a code example. Maybe you can use this in some way for the documentation:
https://github.com/imixs/imixs-bpmn.glsp/blob/master/doc/TOOL_PALETTE.md
